specs/graph-protocol-hybrid-network/read-interface/README.md

```diff
@@ -1,357 +1,273 @@
-- `entitydb` - An [entity database model](#entity-database-model).
-- `rdb` -  Relational database model. Not supported in this version of the protocol.
-- `eav` -  Entity-attribute-value database model. Not supported in this version of the protocol.
+### Example
 
-The database model also defines the available partitions and index types for use in read operations.
-
-In the v1 protocol, we only support the entity database model.
-
-#### Entity Database Model
-In the protocol's entity database model, entities are stored as key-value pairs, where the key is a concatenation of the entity type and the entity ID, and the value is an entity object.
+We assume that we are working with a subgraph that stores musicians, bands and their songs using the following schema:
+```graphql
+type Musician @entity {
+    id: ID!
+    name: String!
+    mainBand: Band
+    bands: [Band!]!
+    writtenSongs: [Song]! @derivedFrom(field: "writtenBy")
+}
 
-This database model is referenced as `entitydb` in Index Records.
+type Band @entity {
+    id: ID!
+    name: String!
+    members: [Musician!]! @derivedFrom(field: "bands")
+    originalSongs: [Song!]!
+}
 
+type Song @entity {
+    id: ID!
+    title: String!
+    writtenBy: Musician!
+    band: Band @derivedFrom(field: "originalSongs")
+}
+```
 
-##### Example - Entities Stored as Key-Value Pairs
-| Key | Value |
-| --- | ----- |
-| `user:1` | `{ id: "user:1", user: "Alice", age: 17 }` |
-| `user:2` | `{ id: "user:2", user: "Bob", age: 47 }` |
+Queries against such a subgraph can make use of the `musicians`, `bands`, and `songs` table. The ID's of the songs a musician has written can be queried with
+```sql
+    select s.id
+    from musicians as m
+    inner join songs as s on s.written_by = m.id
+    where m.name = 'John Coltrane'
+```
 
-#### Partitions
-Partitions define the subset of the data that is covered by the index.
+A GraphQL query that lists musicians, their main bands, and the songs they have written
+```graphql
+    musicians { id mainBand { id } writtenSongs { id } }
+```
 
-Possible values of `<partition>` in an Index Record:
- - none - Includes all entities in the dataset. Default partition, `partition` key should be ommitted in IndexRecord.
- - `"<entityType>"` - Includes entities of the type specified by `entityType`. The entity type name is case-sensitive.
- - `"<interface>"` - Includes entities that implement the provided interface. The interface name is case-sensitive.
+would correspond to this SQL query
+```sql
+    select m.id,
+           map(main_band) as mainBand,
```

Comment:
It seems like the main role that `map` and `collect` are playing here are to explicitly normalize the response data.

Is this strictly necessary or could we just infer that we always want to convert a denormalized set of tabular data into a normalized nested object?

Do you see `map` and `collect` playing an important role in facilitating other GraphQL queries?

Reply:
The main purpose for `map` and `collect` is to change the representation of data, and to let us break out of the strict tabular representation that SQL implies by introducing composite values. Without these, we would be restricted to dealing with tables whose columns can only represent scalar types.

I don't think we can infer that automatically; in the example you quote, the use of `map` and `collect` directly reflects how relationships are modelled in GraphQL; in our subset of SQL, those relationships aren't expressed directly, and `map` and `collect` are used to go from a model that is more flexible in terms of data relationships  to the more restrictive GraphQL model.

Reply:
Okay that rationale makes sense to me. I'm still a little fuzzy, however, on the syntax you're showing here.

You're calling `map` and `collect` in your `select` statement using the aliases that you assigned to the tables you produced during the join.

It seems like the `map` and `collect` should happen "before" the join, while you are producing the tables to be used in the join. By the time you are in the `select` statement shouldnt "songs" and "main_band" just refer to column names? What would it mean to operate on a table in this context?

Maybe something like:
```
    select m.id,
           mainBand.value,
           songs.values
        from musicians m
             left outer join ( map (select id from bands) over id) as main_band
                             on (m.main_band = main_band.key)
             left outer join (collect (select id, written_by from songs) over written_by) as songs
                             on (songs.key = m.id)
```
In the example above I assumed that `map` produces a table with a `key` and `value` column where the `value` column has the result that you described. I assumed that `collect` produces a table with `key` and `values` columns. In both cases `key` is whatever column was passed in to the `over` statement. These might not be the friendliest semantics, but gets the ideas across.

Working through the example I understand better what `collect` is accomplishing, though I'm still having a little trouble with `map`. It seems like it's just formatting that could be done by the query engine? I might need a more complete example to drive it home for me.

Reply:
That's sorta what would happen behind the scenes, but not how you write this stuff in SQL; you can think of the evaluation of a simple `select ... from ... where ...` of first taking the tables in the `from` clause and creating one big result table based on the `where` clause (and join conditions in the `from`) The `select` then talks about which of the columns of that result table to use resp. how to transform columns (in the case of `map`) into new columns.

I _think_ you are right that we could make do without the `map`; I'd need to think more about that and sketch out more examples.